source/common/upstream/transport_socket_matcher.cc

```diff
@@ -0,0 +1,85 @@
+#include "common/upstream/transport_socket_matcher.h"
+
+#include "envoy/server/transport_socket_config.h"
+
+#include "common/config/utility.h"
+
+namespace Envoy {
+namespace Upstream {
+
+TransportSocketMatcher::TransportSocketMatcher(
+    const Protobuf::RepeatedPtrField<envoy::api::v2::Cluster_TransportSocketMatch>& socket_matches,
+    Server::Configuration::TransportSocketFactoryContext& factory_context,
+    Network::TransportSocketFactory& default_factory, Stats::Scope& stats_scope)
+    : default_socket_factory_(default_factory), stats_scope_(stats_scope) {
+  for (const auto& socket_match : socket_matches) {
+    FactoryMatch factory_match(socket_match.name(), generateStats(socket_match.name() + "."));
+    for (const auto& kv : socket_match.match().fields()) {
+      // TODO: question, what's the handling for non string value case?
+      if (kv.second.kind_case() == ProtobufWkt::Value::kStringValue) {
+        factory_match.match[kv.first] = kv.second.string_value();
+      }
+    }
+    const auto& socket_config = socket_match.transport_socket();
+    auto& config_factory = Config::Utility::getAndCheckFactory<
+        Server::Configuration::UpstreamTransportSocketConfigFactory>(socket_config.name());
+    ProtobufTypes::MessagePtr message = Config::Utility::translateToFactoryConfig(
+        socket_config, factory_context.messageValidationVisitor(), config_factory);
+    factory_match.factory = config_factory.createTransportSocketFactory(*message, factory_context);
+    matches_.emplace_back(std::move(factory_match));
+  }
+}
+
+bool metadataMatch(const envoy::api::v2::core::Metadata& metadata,
```

Comment:
Why can't we reuse the matching code that already exists for matching and is part of the subset LB?

Reply:
I think this [`hostMatches`](https://github.com/envoyproxy/envoy/blob/f51fab0d4518e87ae93d71b3c42f8c69d68a9514/source/common/upstream/subset_lb.cc#L437) is very closed to the `metadataMatch` logic here.

I'm thinking maybe we can extract that part of code into a shared library/utility function, used by both `transport_socket_match` and `subset_lb`?

Or you suggest we extend the subset_lb to be able to select socket implementation at all? If so, I feel the `subset_lb` name is not a good fit and mixed with load balancing with transport socket logic?

Reply:
Yeah, that's what I was thinking, some kind of utility to do metadata matching that is shared by both subset LB and the transport socket matcher, similar to `hostMatches()`.

Reply:
done. moved to config/metadata/metadata.cc